src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1501,5 +1497,112 @@ private interface IWebSocketReceiveResultGetter<TResult>
+            private ManualResetValueTaskSourceCore<bool> _mrvts;
+
+            private int _minimumRequiredBytes;
+            private CancellationToken _cancellationToken;
+            private bool _throwOnPrematureClosure;
+
+            private ConfiguredValueTaskAwaitable<int>.ConfiguredValueTaskAwaiter _awaiter;
+
+            public EnsureBufferValueTaskSource(ManagedWebSocket webSocket)
+            {
+                _webSocket = webSocket;
+                _onComplete = new Action(OnComplete);
+            }
+
+            public ValueTask EnsureBufferContainsAsync(int minimumRequiredBytes, CancellationToken cancellationToken, bool throwOnPrematureClosure)
+            {
+                _minimumRequiredBytes = minimumRequiredBytes;
+                _cancellationToken = cancellationToken;
+                _throwOnPrematureClosure = throwOnPrematureClosure;
+
+                return EnsureBufferContains();
+            }
+
+            private ValueTask EnsureBufferContains()
+            {
+                // While we don't have enough data, read more.
+                while (_webSocket._receiveBufferCount < _minimumRequiredBytes)
+                {
+                    ValueTask<int> vt = ReadAsync();
+                    if (!vt.IsCompletedSuccessfully)
+                    {
+                        _awaiter = vt.ConfigureAwait(false).GetAwaiter();
+                        _awaiter.UnsafeOnCompleted(_onComplete);
+                        return new ValueTask(this, _mrvts.Version);
+                    }
+
+                    int numRead = vt.Result;
+                    Debug.Assert(numRead >= 0, $"Expected non-negative bytes read, got {numRead}");
+                    if (numRead <= 0)
+                    {
+                        _webSocket.ThrowIfEOFUnexpected(_throwOnPrematureClosure);
+                        break;
+                    }
+                    _webSocket._receiveBufferCount += numRead;
+                }
+
+                // Completed sync
+                return default;
+            }
+
+            private void OnComplete()
+            {
+                try
+                {
+                    int numRead = _awaiter.GetResult();
+                    Debug.Assert(numRead >= 0, $"Expected non-negative bytes read, got {numRead}");
+                    if (numRead <= 0)
+                    {
+                        _webSocket.ThrowIfEOFUnexpected(_throwOnPrematureClosure);
+                        _mrvts.SetResult(default);
+                        return;
+                    }
+                    _webSocket._receiveBufferCount += numRead;
+
+                    if (_webSocket._receiveBufferCount < _minimumRequiredBytes)
+                    {
+                        ValueTask vt = EnsureBufferContains();
+                        if (vt.IsCompletedSuccessfully)
+                        {
+                            _mrvts.SetResult(default);
+                        }
```

Comment:
This looks like a bug.  Can't this end up racing with EnsureBufferContains completing asynchronously and also trying to call SetResult?

Reply:
Fixed https://github.com/dotnet/corefx/pull/39455/commits/95a182b0cc807c7a708579764639cf1603bc7764

Reply:
Rather https://github.com/dotnet/corefx/pull/39455/commits/2eef2a4dec1342ef59e67f8fe14d899479f3ec32